hooks/hookexecution/execution.go

```diff
@@ -0,0 +1,227 @@
+	rejected <-chan struct{},
+) {
+	hookRespCh := make(chan hookResponse[P], 1)
+	startTime := time.Now()
+	hookId := HookID{ModuleCode: hw.Module, HookCode: hw.Code}
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+		result, err := hookHandler(ctx, moduleCtx, hw.Hook, payload)
+		hookRespCh <- hookResponse[P]{
+			Result: result,
+			Err:    err,
+		}
+	}()
+
+	select {
+	case res := <-hookRespCh:
+		res.HookID = hookId
+		res.ExecutionTime = time.Since(startTime)
+		resp <- res
+	case <-time.After(timeout):
+		resp <- hookResponse[P]{
+			Err:           TimeoutError{},
+			ExecutionTime: time.Since(startTime),
+			HookID:        hookId,
+			Result:        hookstage.HookResult[P]{},
+		}
+	case <-rejected:
+		return
+	}
+}
+
+func collectHookResponses[P any](
+	resp <-chan hookResponse[P],
+	rejected chan<- struct{},
+) []hookResponse[P] {
+	hookResponses := make([]hookResponse[P], 0)
+	for r := range resp {
+		hookResponses = append(hookResponses, r)
+		if r.Result.Reject {
+			close(rejected)
+			break
+		}
+	}
+
+	return hookResponses
+}
+
+func processHookResponses[P any](
+	executionCtx executionContext,
+	hookResponses []hookResponse[P],
+	payload P,
+) (GroupOutcome, P, groupModuleContext, *RejectError) {
+	stage := hooks.Stage(executionCtx.stage)
+	groupOutcome := GroupOutcome{}
+	groupOutcome.InvocationResults = make([]HookOutcome, 0, len(hookResponses))
+	groupModuleCtx := make(groupModuleContext, len(hookResponses))
+
+	for i, r := range hookResponses {
+		groupOutcome.InvocationResults = append(groupOutcome.InvocationResults, HookOutcome{
+			Status:        StatusSuccess,
+			HookID:        r.HookID,
+			Message:       r.Result.Message,
+			DebugMessages: r.Result.DebugMessages,
+			AnalyticsTags: r.Result.AnalyticsTags,
+			ExecutionTime: ExecutionTime{ExecutionTimeMillis: r.ExecutionTime},
+		})
+		groupModuleCtx[r.HookID.ModuleCode] = r.Result.ModuleContext
+
+		if r.ExecutionTime > groupOutcome.ExecutionTimeMillis {
+			groupOutcome.ExecutionTimeMillis = r.ExecutionTime
+		}
+
+		if r.Err != nil {
+			groupOutcome.InvocationResults[i].Errors = append(groupOutcome.InvocationResults[i].Errors, r.Err.Error())
+			switch r.Err.(type) {
+			case TimeoutError:
+				groupOutcome.InvocationResults[i].Status = StatusTimeout
+			case FailureError:
+				groupOutcome.InvocationResults[i].Status = StatusFailure
+			default:
+				groupOutcome.InvocationResults[i].Status = StatusExecutionFailure
+			}
+			continue
+		}
+
+		if r.Result.Reject {
+			if !stage.IsRejectable() {
+				groupOutcome.InvocationResults[i].Status = StatusExecutionFailure
+				groupOutcome.InvocationResults[i].Errors = append(
+					groupOutcome.InvocationResults[i].Errors,
+					fmt.Sprintf(
+						"Module (name: %s, hook code: %s) tried to reject request on the %s stage that does not support rejection",
+						r.HookID.ModuleCode,
+						r.HookID.HookCode,
+						executionCtx.stage,
+					),
```

Comment:
Is it possible to update tests to cover these lines?

Reply:
Unfortunately, right now we have implementation only for the `entrypoint` stage which is rejectable. But we are going to cover this logic with tests within the PRs for `all-processed-bid-responses` and `auction-response` stages, for which rejection is not allowed.